### PR TITLE
[#10156] Add frontend CSV generation framework

### PR DESCRIPTION
**Pull request by xpdavid**
Part of #10156

What is missing:
- Participant comment and Instructor comments
- Students CSV generation

**Commit Organization**

See below

### CODE REVIEW

**Comment by xpdavid** [comment]
I think let's do this migration in this way:
1. We create a new branch on top the `master` and do CSV migration related PRs there. All the review process remains the same.
2. We then merge it back when time comes.

For several reasons:

> There's way more dead code to be removed here.

I actually purposely leave them there for reference.

> Let's not forget to also move the CSV generation of students list here

It might be too much to do this in the PR :P

> Unless you plan to do the full migration in this PR, merging this to master is not advisable since this is taking away a working feature.

That's why I suggest this way.

> This is not a good way. Above all, this adds a tight coupling between front-end testing and back-end testing

Yup. It's bad. My thought is that somehow the backend will change (most likely during the migration) and it is hard to update the data file manually.

> This has no meaning once the mentioned CSV file is removed. Let's not add this kind of line.

The severs as reference purpose during the migration (e.g., check for correctness).

File: src/web/services/csv-helper.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/**
+ * CSV related utility functions.
+ */
+export class CsvHelper {
```
[the remark this answers is not in the document]
Comment (xpdavid) [reply src/web/services/csv-helper.ts:4]: Those libraries provide high-level conversion between JOSN and CSV but our results CSV generations requires some low-level manipulations. It is worth to try especially for the student list generation.

File: src/web/services/session-result-csv.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+import { SessionResultCsvService } from './session-result-csv.service';
+
+/**
+ * Loads data for testing.
+ */
+const loadTestData: (filename: string) => SessionResults = (filename: string): SessionResults => {
+  try {
+    return require(`../../test/resources/generated-data/${filename}`);
+  } catch (error) {
+    console.error('Fail to load test data, did you run GenerateSessionResults?');
```
[the remark this answers is not in the document]
Comment (xpdavid) [reply src/web/services/session-result-csv.service.spec.ts]: Included generated files instead. I believe the API output would be almost fixed. Let's revise the test file (to a single generic-cover-all test file) in another PR.

File: src/web/services/session-result-csv.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    expect(service).toBeTruthy();
+  });
+
+  it('should show responses for feedbackSessionResultsC1S1', () => {
+    const sessionResult: SessionResults = loadTestData('feedbackSessionResultsC1S1.json');
+
+    const result: string = service.getCsvForSessionResult(sessionResult,
+        false, false);
+    expect(replaceUnpredictableValuesWithPlaceholders(result)).toMatchSnapshot();
+    // feedbackSessionResultsC1S1_actionTest.csv
```
[the remark this answers is not in the document]
Comment (xpdavid) [reply src/web/services/session-result-csv.service.spec.ts:63]: Choose to leave them there for migration purpose. Will remove them when the branch is ready to be merged.

File: src/web/services/csv-helper.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/web/services/csv-helper.ts:4]: Then for now, at least you can instead make use of this method to expedite many of the processes:
```ts
convertCsvContentsToCsvString(rows: string[][]): string {
  return rows.map((columns: string[]) => columns.map((entry: string) => {
    if (entry.indexOf(',') >= 0 || entry.indexOf('"') >= 0) {
      return `"${entry.replace(/"/g, '""')}"`;
    }
    return entry;
  }).join(',')).join('\n');
}
```

File: src/web/types/question-details-impl/feedback-question-details-factory.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import { AbstractFeedbackQuestionDetails } from './abstract-feedback-question-details';
+import { FeedbackConstantSumOptionsQuestionDetailsImpl } from './feedback-constsum-options-question-details.impl';
+import { FeedbackConstantSumRecipientsQuestionDetailsImpl } from './feedback-constsum-recipient-question-details.impl';
+import { FeedbackContributionQuestionDetailsImpl } from './feedback-contribution-question-details.impl';
+import { FeedbackMcqQuestionDetailsImpl } from './feedback-mcq-question-details.impl';
+import { FeedbackMsqQuestionDetailsImpl } from './feedback-msq-question-details.impl';
+import { FeedbackNumericalScaleQuestionDetailsImpl } from './feedback-num-scale-question-details.impl';
+import { FeedbackRankOptionsQuestionDetailsImpl } from './feedback-rank-options-question-details.impl';
+import { FeedbackRankRecipientsQuestionDetailsImpl } from './feedback-rank-recipients-question-details.impl';
+import { FeedbackRubricQuestionDetailsImpl } from './feedback-rubric-quesiton-details.impl';
```
Comment (wkurniawan07) [inline src/web/types/question-details-impl/feedback-question-details-factory.ts:23]: Typo in the file name `quesiton`

**Comment by xpdavid** [comment]
What's new:

- Add stats for all question types for CSV generation

What's missing:

- Participant/Instructor Comment
- Student List CSV generation

What's fixed:
The fixes are to align with the old CSV generation.

- MSQ/MCQ: Fix wrong weight average calculation
Before:
https://github.com/TEAMMATES/teammates/blob/3ee4ac99450080d094f50939785fa82eea34389f/src/web/app/components/question-types/question-statistics/mcq-question-statistics.component.ts#L84-L85
- Rank: Fix wrong `rankPerOption` calculation
Before:
https://github.com/TEAMMATES/teammates/blob/3ee4ac99450080d094f50939785fa82eea34389f/src/web/app/components/question-types/question-statistics/rank-options-question-statistics.component.ts#L63
Assigning `0` to options with no response will always result in the option to be the first
- Rubric: Add missing feature: Per recipient stats
image
- Rubric: Add missing feature: Average weights
image
- Rubric: Fix wrong weight percentage calculation. The percentage should not be weighted
Before:
https://github.com/TEAMMATES/teammates/blob/3ee4ac99450080d094f50939785fa82eea34389f/src/web/app/components/question-types/question-statistics/rubric-question-statistics.component.ts#L99-L106

File: src/web/app/components/question-types/question-statistics/constsum-options-question-statistics.component.ts [hunk 5]
```diff
@@ -1,9 +1,10 @@
 import { Component, OnChanges, OnInit } from '@angular/core';
-import { FeedbackConstantSumQuestionDetails, FeedbackConstantSumResponseDetails } from '../../../../types/api-output';
 import { DEFAULT_CONSTSUM_RECIPIENTS_QUESTION_DETAILS } from '../../../../types/default-question-structs';
 import { SortBy } from '../../../../types/sort-properties';
 import { ColumnData, SortableTableCellData } from '../../sortable-table/sortable-table.component';
-import { QuestionStatistics } from './question-statistics';
+import {
+  ConstsumOptionsQuestionStatisticsCalculation,
+} from './question-statistics-calculation/constsum-options-question-statstics-calculation';
```
Comment (wkurniawan07) [inline src/web/app/components/question-types/question-statistics/constsum-options-question-statistics.component.ts:7]: Typo in filename `statstics`

**Comment by xpdavid** [comment]
What's new:
- Fix according to the review comments
- Update stats response filtering logic to follow #10211
https://github.com/TEAMMATES/teammates/pull/10157/commits/4d45e367a52200cca9348a95df335ade4b358754#diff-59ac34c0bf3b3a2313ae40066dca4163R41-R44

**Pull request merged**